core/src/main/java/org/apache/druid/segment/column/ValueType.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.column;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import org.apache.druid.java.util.common.StringUtils;
+
+import javax.annotation.Nullable;
+
+public enum ValueType
+{
+  DOUBLE,
+  FLOAT,
+  LONG,
+  STRING,
+  COMPLEX,
+  DOUBLE_ARRAY,
+  LONG_ARRAY,
+  STRING_ARRAY;
```

Comment:
Please add javadocs documenting these types.

Especially important things to document include:

- When do we use `STRING` and when do we use `STRING_ARRAY`? (Multivalue strings are type STRING, even though they behave kind of like arrays. How do we explain this coherently?)
- What does COMPLEX mean and how can you get more information about something that is COMPLEX?

Reply:
I am still working on javadocs here and other modified places to provide guidance on how types are and should be used, and how to deal with array types until they are fully propagated through the codebase. 

With regards to `COMPLEX`, I'm starting to think that in the future we might want to modify `RowSignature` to potentially have the ability to store the complex type name, similar to what is available for aggregator factory, and what #10277 adds to the `ColumnCapabilities` of complex type columns it creates. We might actually want a lighter weight `ColumnSignature` type that is a subset of `ColumnCapabilities` that `RowSignature` can be composed of, rather than it's strict `ValueType` mapping it currently has, but I'm not certain if it is necessary yet.

Reply:
will it be too complicated for multi value strings to have their own type different from single value string?

Reply:
Not sure yet, I'm still thinking on the best way to live in a world with the current `STRING` that can have single or multi values, communicated through `ColumnCapabilities.hasMultipleValues`, and `STRING_ARRAY` which is explicitly always multi-valued but can currently only be produced via expressions. 

I think it depends on how we want to encode this information for `RowSignature` to make available to the broker and higher layers of the query engines. There might be room for a new `ValueType` to use explicitly for `STRING` which are multi-valued if we want to keep `RowSignature` light (and effectively coerce it back to `STRING` when translating the signature back into `ColumnCapabilities` for things like the row selectors the broker uses), though between the changes in #10219 which also adds a want to be able to encode in the `RowSignature` which columns can have null values, making a richer `RowSignature` is probably the right way forward, which could potentially make a separate `ValueType` for multi-value strings not necessary.

I don't think we want to treat the multi-value strings as `STRING_ARRAY` because I think we probably want to reserve it for if/when we add true array typed columns, so that engines like group by and top-n can process them separately than the funny way we handle existing multi-value strings (which aggregate on individual values, basically  equivalent to `UNNEST` in SQL) and instead do it in a way that is compatible with SQL array types.

Reply:
I've added a bunch of javadocs trying to advise on what each ValueType represents and how it may be used in the query engines, I hope it is sufficient to provide guidance for complex type/agg/post-agg implementors to think about how the thing is going to be used, but I'm sure that I'm probably leaving out a lot of details, and it might be somewhat brittle to keep in sync with reality as the engines evolve.